src/symbolize/gimli/elf.rs

```diff
@@ -43,6 +43,45 @@ impl Mapping {
         })
     }
 
+    /// On Android, shared objects can be loaded directly from a
+    /// ZIP archive. For example, an app may load a library from
+    /// `/data/app/com.example/base.apk!/lib/x86_64/mylib.so`
+    ///
+    /// `zip_offset` should be page-aligned; the dynamic linker
+    /// requires this when it loads libraries.
+    #[cfg(target_os = "android")]
+    pub fn new_android(path: &Path, zip_offset: Option<i64>) -> Option<Mapping> {
+        fn map_embedded_library(path: &Path, zip_offset: i64) -> Option<Mapping> {
+            // get path of ZIP archive (delimited by `!/`)
+            let raw_path = path.as_os_str().as_bytes();
+            let zip_path = raw_path
+                .windows(2)
+                .enumerate()
+                .find(|(_, chunk)| chunk == b"!/")
+                .map(|(index, _)| Path::new(OsStr::from_bytes(raw_path.split_at(index).0)))?;
```

Comment:
...so... does Android actually reject including `!` as the final character in a filename on the OS level? Is there no way to create such a file?

Reply:
It does not prevent that - you can even have both paths existing simultaneously (e.g. `/path/to/my!/file` as a file within a directory at `/path/to/my!` *and* as a library in an archive named `/path/to/my`).

Maybe I should document this more clearly, but that's why the code first tries loading from an archive, and then falls back to trying to load the path literally.
The embedded path takes precedence to match [the behaviour of Android's dynamic linker](https://cs.android.com/android/platform/superproject/main/+/daa7e5ca6b9d2ae543e60be31502c302d08304e1:bionic/linker/linker.cpp;l=998), which has the same fallback behaviour.